### PR TITLE
Remove ILambda user tree node from Painless

### LINKED ISSUE #54015: Remove remaining mutable state in Painless user tree
The following are a few places in the Painless user tree that still contain mutable state:

1. PBrace and PField with the isDefOptimized variable
2. nodes extending ILambda with the ref variable used to create a "pointer" (info about the lambda for use during runtime) in PSubDefCall

We will work to remove these with further refactoring.

### PR DESCRIPTION
**Pull request by jdconrad**
The ILambda user tree node is no longer necessary with the addition of the ir tree as it was only supporting writing the ASM, but had no bearing on semantic checking. This moves the code necessary to build the def lambda recipes into the ir tree since we know that they are correct for compile time and the information necessary to do this is already known by the correct ir nodes.

This change also splits up the ir reference nodes into DefReferenceInterfaceNode, TypedInterfaceReferenceNode, and TypedCaptureReferenceNode which covers all possible cases of known information at compile time for method references. Each of these nodes is built by the user tree nodes as necessary.

Relates #53702
Closes #54015

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java [hunk 1]
```diff
@@ -45,50 +44,49 @@ public String getName() {
...
-    }
+        StringBuilder defCallRecipe = new StringBuilder();
+        List<Object> boostrapArguments = new ArrayList<>();
+        List<Class<?>> typeParameters = new ArrayList<>();
+        int capturedCount = 0;
 
-    public List<String> getPointers() {
-        return pointers;
-    }
+        typeParameters.add(Object.class);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:58]: Reciever?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java [hunk 2]
```diff
@@ -45,50 +44,49 @@ public String getName() {
...
+    protected void write(ClassWriter classWriter, MethodWriter methodWriter, ScopeTable scopeTable) {
+        methodWriter.writeDebugInfo(location);
 
-    public void addPointer(String pointer) {
-        pointers.add(pointer);
-    }
+        StringBuilder defCallRecipe = new StringBuilder();
+        List<Object> boostrapArguments = new ArrayList<>();
+        List<Class<?>> typeParameters = new ArrayList<>();
+        int capturedCount = 0;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:56]: ?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java [hunk 3]
```diff
@@ -45,50 +44,49 @@ public String getName() {
...
 
-    /* ---- end node data ---- */
+            if (argumentNode instanceof DefInterfaceReferenceNode) {
+                DefInterfaceReferenceNode defInterfaceReferenceNode = (DefInterfaceReferenceNode)argumentNode;
+                boostrapArguments.add(defInterfaceReferenceNode.getDefReferenceEncoding());
 
-    @Override
-    protected void write(ClassWriter classWriter, MethodWriter methodWriter, ScopeTable scopeTable) {
-        methodWriter.writeDebugInfo(location);
+                char encoding = (char)(i + capturedCount);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:70]: ?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DefInterfaceReferenceNode.java [hunk 4]
```diff
@@ -20,34 +20,36 @@
...
     protected void write(ClassWriter classWriter, MethodWriter methodWriter, ScopeTable scopeTable) {
-        if (funcRef != null) {
-            methodWriter.writeDebugInfo(location);
-            methodWriter.invokeLambdaCall(funcRef);
-        } else {
-            // push a null instruction as a placeholder for future lambda instructions
-            methodWriter.push((String)null);
+        methodWriter.writeDebugInfo(location);
+
+        // place holder for
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DefInterfaceReferenceNode.java:47]: don't leave me hanging!

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DefInterfaceReferenceNode.java [hunk 5]
```diff
@@ -20,34 +20,36 @@
...
+        methodWriter.push((String)null);
+
+        for (String capture : getCaptures()) {
+            ScopeTable.Variable variable = scopeTable.getVariable(capture);
+            methodWriter.visitVarInsn(variable.getAsmType().getOpcode(Opcodes.ILOAD), variable.getSlot());
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DefInterfaceReferenceNode.java:52]: `visitVarInsn` -> visit variable instance?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ReferenceNode.java [hunk 6]
```diff
@@ -7,7 +7,7 @@
  * not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *    http://www.apache.org/licenses/LICENSE-2.0
+ *     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ReferenceNode.java:10]: ?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ReferenceNode.java [hunk 7]
```diff
@@ -17,26 +17,25 @@
...
-    List<Class<?>> getCaptures();
+    public void addCapture(String capture) {
+        captures.add(capture);
+    }
 
-    /** Returns the number of captured parameters */
-    default int getCaptureCount() {
-        return getCaptures().size();
+    public List<String> getCaptures() {
+        return captures;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ReferenceNode.java:36]: is it ok to leak a mutable data structure here?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/TypedCaptureReferenceNode.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    public String getMethodName() {
+        return methodName;
+    }
+
+    /* ---- end node data ---- */
+
+    @Override
+    protected void write(ClassWriter classWriter, MethodWriter methodWriter, ScopeTable scopeTable) {
+        methodWriter.writeDebugInfo(location);
+        Variable captured = scopeTable.getVariable(getCaptures().get(0));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/TypedCaptureReferenceNode.java:49]: why is the `get(0)` safe here?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java [hunk 9]
```diff
@@ -45,50 +44,49 @@ public String getName() {
...
 
-    public void setRecipe(String recipe) {
-        this.recipe = recipe;
-    }
+    /* ---- end node data ---- */
 
-    public String getRecipe() {
-        return recipe;
-    }
+    @Override
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:49]: Javadocs: "Write out invokedynamic instruction" or something similar.  Nothing fancy.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java [hunk 10]
```diff
@@ -45,50 +44,49 @@ public String getName() {
...
+    protected void write(ClassWriter classWriter, MethodWriter methodWriter, ScopeTable scopeTable) {
+        methodWriter.writeDebugInfo(location);
 
-    public void addPointer(String pointer) {
-        pointers.add(pointer);
-    }
+        StringBuilder defCallRecipe = new StringBuilder();
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:53]: "Handle lambdas as args"

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java [hunk 11]
```diff
@@ -45,50 +44,49 @@ public String getName() {
...
+            ExpressionNode argumentNode = getArgumentNodes().get(i);
+            argumentNode.write(classWriter, methodWriter, scopeTable);
 
-    public List<Class<?>> getTypeParameters() {
-        return typeParameters;
-    }
+            typeParameters.add(argumentNode.getExpressionType());
 
-    /* ---- end node data ---- */
+            if (argumentNode instanceof DefInterfaceReferenceNode) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:66]: This is an unknown method ref that we have to resolve at runtime.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java (hunk 3 above)
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:70]: Small comment about the char encoding skip table, perhaps refing some other documentation, would be helpful.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java [hunk 12]
```diff
@@ -63,51 +59,53 @@ Output analyze(ClassNode classNode, ScriptRoot scriptRoot, Scope scope, Input in
...
         Output output = new Output();
 
-        captured = scope.getVariable(location, variable);
+        Variable captured = scope.getVariable(location, variable);
         if (input.expected == null) {
+            String defReferenceEncoding;
             if (captured.getType() == def.class) {
                 // dynamic implementation
-                defPointer = "D" + variable + "." + call + ",1";
+                defReferenceEncoding = "D" + variable + "." + call + ",1";
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java:69]: Add TODO, let's pull this encoding creation out into something that makes the arguments clear.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java [hunk 13]
```diff
@@ -63,51 +59,53 @@ Output analyze(ClassNode classNode, ScriptRoot scriptRoot, Scope scope, Input in
...
 
-        capturingFuncRefNode.setLocation(location);
-        capturingFuncRefNode.setExpressionType(output.actual);
-        capturingFuncRefNode.setCapturedName(captured.getName());
-        capturingFuncRefNode.setName(call);
-        capturingFuncRefNode.setPointer(defPointer);
-        capturingFuncRefNode.setFuncRef(ref);
+                output.expressionNode = typedInterfaceReferenceNode;
+            } else {
+                TypedCaptureReferenceNode typedCaptureReferenceNode = new TypedCaptureReferenceNode();
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java:99]: Add comment on this case.

**Review by stu-elastic — APPROVED** [review]
Keep on truckin!

**Comment by jdconrad** [comment]
@stu-elastic Thank you for the feedback. I will address all of your comments soon.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:58]: commented

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:56]: commented

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:70]: commented

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DefInterfaceReferenceNode.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DefInterfaceReferenceNode.java:52]: an asm instruction to load a variable's value onto the stack

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ReferenceNode.java (hunk 7 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ReferenceNode.java:36]: all ir nodes are designed to be completely mutable as we hope to modify the ir tree structure at some point to add optimizations and possibly other features; we keep the user tree as immutable and hide the ir tree behind it

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/TypedCaptureReferenceNode.java (hunk 8 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/TypedCaptureReferenceNode.java:49]: there is only a single capture for reference types in this case (foo::bar where foo is the captured variable)

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java (hunk 10 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:53]: commented

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java (hunk 9 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:49]: commented

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java (hunk 11 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:66]: commented

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DefInterfaceReferenceNode.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DefInterfaceReferenceNode.java:47]: oops, fixed :)

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java (hunk 12 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java:69]: https://github.com/elastic/elasticsearch/issues/55543

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java (hunk 13 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java:99]: commented

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/CallSubDefNode.java:70]: commented

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-unix-docker

**Pull request merged**